File: scala/scalasearch/src/main/scala/scalasearch/SearchOptions.scala
package scalasearch

import scala.collection.mutable
import scala.xml.XML

case class SearchOption(shortarg:String, longarg:String, desc:String) {
  val sortarg =
    if (shortarg.nonEmpty) {
      shortarg.toLowerCase + "@" + longarg.toLowerCase
    } else {
      longarg.toLowerCase
    }
}

object SearchOptions {
  // TODO: move to config file
  private val _searchOptionsXmlPath = "/searchoptions.xml"
  private val _searchOptions = mutable.ListBuffer.empty[SearchOption]

  private def searchOptions: List[SearchOption] = {
    if (_searchOptions.isEmpty) {
      val root = XML.load(getClass.getResourceAsStream(_searchOptionsXmlPath))
      val searchOptionNodes = root \\ "searchoption"
      for (searchOptionNode <- searchOptionNodes) {
        val short = (searchOptionNode \ "@short").text
        val long = (searchOptionNode \ "@long").text
        val desc = searchOptionNode.text.trim
        val option = SearchOption(short, long, desc)
        _searchOptions += option
      }
    }
    List.empty[SearchOption] ++ _searchOptions.sortWith(_.sortarg < _.sortarg)
  }

  val argActionMap = Map[String, ((String, SettingsBuilder) => Unit)](
    "in-archiveext" ->
      ((x: String, sb: SettingsBuilder) => sb.addInArchiveExtensions(x)),
    "in-archivefilepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addInArchiveFilePattern(x)),
    "in-dirpattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addInDirPattern(x)),
    "in-ext" ->
      ((x: String, sb: SettingsBuilder) => sb.addInExtensions(x)),
    "in-filepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addInFilePattern(x)),
    "in-linesafterpattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addInLinesAfterPattern(x)),
    "in-linesbeforepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addInLinesBeforePattern(x)),
    "linesafter" ->
      ((x: String, sb: SettingsBuilder) => sb.linesAfter = x.toInt),
    "linesaftertopattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addLinesAfterToPattern(x)),
    "linesafteruntilpattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addLinesAfterUntilPattern(x)),
    "linesbefore" ->
      ((x: String, sb: SettingsBuilder) => sb.linesBefore = x.toInt),
    "maxlinelength" ->
      ((x: String, sb: SettingsBuilder) => sb.maxLineLength = x.toInt),
    "out-archiveext" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutArchiveExtensions(x)),
    "out-archivefilepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutArchiveFilePattern(x)),
    "out-dirpattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutDirPattern(x)),
    "out-ext" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutExtensions(x)),
    "out-filepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutFilePattern(x)),
    "out-linesafterpattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutLinesAfterPattern(x)),
    "out-linesbeforepattern" ->
      ((x: String, sb: SettingsBuilder) => sb.addOutLinesBeforePattern(x)),
    "search" ->
      ((x: String, sb: SettingsBuilder) => sb.addSearchPattern(x))
  )

  val flagActionMap = Map[String, (SettingsBuilder => Unit)](
    "archivesonly" ->
      ((sb: SettingsBuilder) => sb.setArchivesOnly()),
    "allmatches" ->
      ((sb: SettingsBuilder) => sb.firstMatch = false),
    "debug" ->
      ((sb: SettingsBuilder) => sb.setDebug()),
    "excludehidden" ->
      ((sb: SettingsBuilder) => sb.excludeHidden = true),
    "firstmatch" ->
      ((sb: SettingsBuilder) => sb.firstMatch = true),
    "help" ->
      ((sb: SettingsBuilder) => sb.printUsage = true),
    "includehidden" ->
      ((sb: SettingsBuilder) => sb.excludeHidden = false),
    "listdirs" ->
      ((sb: SettingsBuilder) => sb.listDirs = true),
    "listfiles" ->
      ((sb: SettingsBuilder) => sb.listFiles = true),
    "listlines" ->
      ((sb: SettingsBuilder) => sb.listLines = true),
    "multilinesearch" ->
      ((sb: SettingsBuilder) => sb.multiLineSearch = true),
    "noprintmatches" ->
      ((sb: SettingsBuilder) => sb.printResults = false),
    "norecursive" ->
      ((sb: SettingsBuilder) => sb.recursive = false),
    "nosearcharchives" ->
      ((sb: SettingsBuilder) => sb.searchArchives = false),
    "printmatches" ->
      ((sb: SettingsBuilder) => sb.printResults = true),
    "recursive" ->
      ((sb: SettingsBuilder) => sb.recursive = true),
    "searcharchives" ->
      ((sb: SettingsBuilder) => sb.searchArchives = true),
    "uniquelines" ->
      ((sb: SettingsBuilder) => sb.uniqueLines = true),
    "verbose" ->
      ((sb: SettingsBuilder) => sb.verbose = true),
    "version" ->
      ((sb: SettingsBuilder) => sb.printVersion = true)
  )

  private def mapFromOptions(options: List[SearchOption]): Map[String,SearchOption] = {
    (options.map(o => (o.longarg, o)) ++
      options.filter(o => o.shortarg.length > 0).map(o => (o.shortarg, o))).toMap
  }

  def settingsFromArgs(args: Array[String]): SearchSettings = {
    val sb = new SettingsBuilder
    val argMap = mapFromOptions(searchOptions.filter(o => argActionMap.contains(o.longarg)))
    val flagMap = mapFromOptions(searchOptions.filter(o => flagActionMap.contains(o.longarg)))
    val switchPattern = """^\-+(\w[\w\-]*)$""".r
    def nextArg(arglist:List[String], sb:SettingsBuilder) {
      arglist match {
        case Nil =>
        case switchPattern(arg) :: tail =>
          if (argMap.contains(arg)) {
            if (tail.nonEmpty) {
              argActionMap(argMap(arg).longarg)(tail.head, sb)
              nextArg(tail.tail, sb)
            } else {
              throw new SearchException("Missing value for arg %s".format(arg))
            }
          } else if (flagMap.contains(arg)) {
            flagActionMap(flagMap(arg).longarg)(sb)
            nextArg(tail, sb)
          } else {
            throw new SearchException("Unknown option: %s".format(arg))
          }
        case arg :: tail =>
          sb.startPath = Some(arg)
          nextArg(tail, sb)
      }
    }
    nextArg(args.toList, sb)
    sb.toSettings
  }

  def usage(status: Int): Unit = {
    Common.log(getUsageString)
    sys.exit(status)
  }

  def getUsageString: String = {
    val sb = new StringBuilder
    sb.append("Usage:\n")
    sb.append(" scalasearch [options] -s <searchpattern> <startpath>\n\n")
    sb.append("Options:\n")
    val optStrings =
      (searchOptions.map(o => if (o.shortarg.nonEmpty) "-" + o.shortarg + "," else "")
        zip searchOptions.map("--" + _.longarg)).map(o => o._1 + o._2)
    val optDescs = searchOptions.map(_.desc)
    val longest = optStrings.map(_.length).max
    val format = " %1$-" + longest + "s  %2$s\n"
    for (i <- optStrings.indices) {
      sb.append(format.format(optStrings(i), optDescs(i)))
    }
    sb.toString()
  }
}

File: fsharp/FsSearch/SearchOptions.fs
﻿namespace FsSearch

open System
open System.Collections.Generic
open System.IO
open System.Text
open System.Text.RegularExpressions
open System.Xml.Linq

type SearchSettings = Settings.SearchSettings

module SearchOptions =   

    type SearchOption = {
        ShortArg : string;
        LongArg : string;
        Description : string
    }

    let addExtensions (exts : string) (extList : string list) : string list =
        List.append extList (FileUtil.ExtensionsListFromString exts)

    let argActionMap : Map<string, string -> SearchSettings -> SearchSettings> =
        [
            ("in-archiveext", (fun (s : string) (settings : SearchSettings) -> { settings with InArchiveExtensions = addExtensions s settings.InArchiveExtensions }));
            ("in-archivefilepattern", (fun (s : string) (settings : SearchSettings) -> { settings with InArchiveFilePatterns = List.append settings.InArchiveFilePatterns [new Regex(s)] }));
            ("in-dirpattern", (fun (s : string) (settings : SearchSettings) -> { settings with InDirPatterns = List.append settings.InDirPatterns [new Regex(s)] }));
            ("in-ext", (fun (s : string) (settings : SearchSettings) -> { settings with InExtensions = addExtensions s settings.InExtensions }));
            ("in-filepattern", (fun (s : string) (settings : SearchSettings) -> { settings with InFilePatterns = List.append settings.InFilePatterns [new Regex(s)] }));
            ("in-linesafterpattern", (fun (s : string) (settings : SearchSettings) -> { settings with InLinesAfterPatterns = List.append settings.InLinesAfterPatterns [new Regex(s)] }));
            ("in-linesbeforepattern", (fun (s : string) (settings : SearchSettings) -> { settings with InLinesBeforePatterns = List.append settings.InLinesBeforePatterns [new Regex(s)] }));
            ("linesafter", (fun (s : string) (settings : SearchSettings) -> { settings with LinesAfter = Int32.Parse(s) }));
            ("linesaftertopattern", (fun (s : string) (settings : SearchSettings) -> { settings with LinesAfterToPatterns = List.append settings.LinesAfterToPatterns [new Regex(s)] }));
            ("linesafteruntilpattern", (fun (s : string) (settings : SearchSettings) -> { settings with LinesAfterUntilPatterns = List.append settings.LinesAfterUntilPatterns [new Regex(s)] }));
            ("linesbefore", (fun (s : string) (settings : SearchSettings) -> { settings with LinesBefore = Int32.Parse(s) }));
            ("maxlinelength", (fun (s : string) (settings : SearchSettings) -> { settings with MaxLineLength = Int32.Parse(s) }));
            ("out-archiveext", (fun (s : string) (settings : SearchSettings) -> { settings with OutArchiveExtensions = addExtensions s settings.OutArchiveExtensions }));
            ("out-archivefilepattern", (fun (s : string) (settings : SearchSettings) -> { settings with OutArchiveFilePatterns = List.append settings.OutArchiveFilePatterns [new Regex(s)] }));
            ("out-dirpattern", (fun (s : string) (settings : SearchSettings) -> { settings with OutDirPatterns = List.append settings.OutDirPatterns [new Regex(s)] }));
            ("out-ext", (fun (s : string) (settings : SearchSettings) -> { settings with OutExtensions = addExtensions s settings.OutExtensions }));
            ("out-filepattern", (fun (s : string) (settings : SearchSettings) -> { settings with OutFilePatterns = List.append settings.OutFilePatterns [new Regex(s)] }));
            ("out-linesafterpattern", (fun (s : string) (settings : SearchSettings) -> { settings with OutLinesAfterPatterns = List.append settings.OutLinesAfterPatterns [new Regex(s)] }));
            ("out-linesbeforepattern", (fun (s : string) (settings : SearchSettings) -> { settings with OutLinesBeforePatterns = List.append settings.OutLinesBeforePatterns [new Regex(s)] }));
            ("search", (fun (s : string) (settings : SearchSettings) -> { settings with SearchPatterns = List.append settings.SearchPatterns [new Regex(s)] }));
        ] |> Map.ofList

    let flagActionMap : Map<string, bool -> SearchSettings -> SearchSettings> =
        [
            ("allmatches", (fun (b : bool) (settings : SearchSettings) -> { settings with FirstMatch = not b }));
            ("archivesonly", (fun (b : bool) (settings : SearchSettings) -> { settings with ArchivesOnly = b }));
            ("debug", (fun (b : bool) (settings : SearchSettings) -> { settings with Debug = true; Verbose = b }));
            ("excludehidden", (fun (b : bool) (settings : SearchSettings) -> { settings with ExcludeHidden = b }));
            ("firstmatch", (fun (b : bool) (settings : SearchSettings) -> { settings with FirstMatch = b }));
            ("help", (fun (b : bool) (settings : SearchSettings) -> { settings with PrintUsage = b }));
            ("includehidden", (fun (b : bool) (settings : SearchSettings) -> { settings with ExcludeHidden = not b }));
            ("listdirs", (fun (b : bool) (settings : SearchSettings) -> { settings with ListDirs = b }));
            ("listfiles", (fun (b : bool) (settings : SearchSettings) -> { settings with ListFiles = b }));
            ("listlines", (fun (b : bool) (settings : SearchSettings) -> { settings with ListLines = b }));
            ("multilinesearch", (fun (b : bool) (settings : SearchSettings) -> { settings with MultiLineSearch = b }));
            ("noprintmatches", (fun (b : bool) (settings : SearchSettings) -> { settings with PrintResults = not b }));
            ("norecursive", (fun (b : bool) (settings : SearchSettings) -> { settings with Recursive = not b }));
            ("nosearcharchives", (fun (b : bool) (settings : SearchSettings) -> { settings with SearchArchives = not b }));
            ("printmatches", (fun (b : bool) (settings : SearchSettings) -> { settings with PrintResults = b }));
            ("recursive", (fun (b : bool) (settings : SearchSettings) -> { settings with Recursive = b }));
            ("searcharchives", (fun (b : bool) (settings : SearchSettings) -> { settings with SearchArchives = b }));
            ("uniquelines", (fun (b : bool) (settings : SearchSettings) -> { settings with UniqueLines = b }));
            ("verbose", (fun (b : bool) (settings : SearchSettings) -> { settings with Verbose = b }));
            ("version", (fun (b : bool) (settings : SearchSettings) -> { settings with PrintVersion = b }));
        ] |> Map.ofList;

    let OptionsFromXml () : SearchOption list =
        let rec recOptionsFromXml (nodeList : XElement list) (options : SearchOption list) : SearchOption list =
            match nodeList with
            | [] -> options
            | n :: ns ->
                let short = [for a in n.Attributes(XName.Get("short")) do yield a.Value].Head
                let long = [for a in n.Attributes(XName.Get("long")) do yield a.Value].Head
                let desc = n.Value.Trim()
                recOptionsFromXml ns (List.append options [{ ShortArg=short; LongArg=long; Description=desc }])
        let _searchOptionsPath = Path.Combine(Config.XSEARCHPATH, "shared/searchoptions.xml")
        let _fileStream = new FileStream(FileUtil.ExpandPath(_searchOptionsPath), FileMode.Open)
        let optNodes = XDocument.Load(_fileStream).Descendants(XName.Get("searchoption"))
        recOptionsFromXml (List.ofSeq optNodes) []
        |> List.sortBy (fun o -> if (o.ShortArg <> "") then (o.ShortArg.ToLower() + "@" + o.LongArg) else o.LongArg)

    let options = OptionsFromXml()

    let SettingsFromArgs (args : string[]) : SearchSettings * string =
        let optionNameMap =
            let shortargs = seq { for opt in options do if opt.ShortArg <> "" then yield (opt.ShortArg, opt.LongArg) }
            let longargs =  seq { for opt in options do yield (opt.LongArg, opt.LongArg) }
            Seq.append shortargs longargs
            |> Map.ofSeq

        let argRegex = new Regex("^(?:-{1,2})(?<opt>.*)$")
        
        let (|IsOption|_|) (arg:string) =            
            let m = argRegex.Match(arg)
            if m.Success then Some(m.Groups.["opt"].Value) else None

        let rec recSettingsFromArgs (argList : string list) (settings : SearchSettings) : SearchSettings * string =
            match argList with
            | [] -> settings, ""
            | head :: tail ->
                match head with
                | IsOption opt ->
                    if (optionNameMap.ContainsKey(opt)) then
                        let long = optionNameMap.[opt]
                        if (argActionMap.ContainsKey(long)) then
                            match tail with
                            | [] ->
                                settings, sprintf "Missing value for option: %s" opt
                            | aHead :: aTail -> 
                                //argActionMap.[long] aHead settings
                                recSettingsFromArgs aTail (argActionMap.[long] aHead settings)
                        elif (flagActionMap.ContainsKey(long)) then
                            //flagActionMap.[long] settings
                            recSettingsFromArgs tail (flagActionMap.[long] true settings)
                        else
                            settings, sprintf "Invalid option: %s" opt
                    else
                        settings, sprintf "Invalid option: %s" opt
                | _ -> recSettingsFromArgs tail { settings with StartPath = head }
        recSettingsFromArgs (Array.toList args) { Settings.DefaultSettings with PrintResults = true }

    let GetUsageString () : string =
        let optStringMap =
            [ for opt in options do
                let shortstring : string = 
                    if opt.ShortArg <> "" then "-" + opt.ShortArg + ","
                    else ""
                let longstring : string = "--" + opt.LongArg
                yield (opt.LongArg, shortstring + longstring) ]
            |> Map.ofList

        let optDescMap =
            [ for opt in options do
                yield (opt.LongArg, opt.Description) ]
            |> Map.ofList

        let optStrings : string list =
            optStringMap
            |> Map.toList
            |> List.map snd

        let longest = 
            optStrings
            |> Seq.map (fun o -> o.Length)
            |> Seq.max

        let format = " {0,-" + longest.ToString() + "}  {1}"

        let usageStrings =
            [for o in options do
                yield String.Format(format, optStringMap.[o.LongArg], optDescMap.[o.LongArg])]

        let usageString = 
            usageStrings
            |> List.append ["\nUsage:"; " FsSearch.exe [options] -s <searchpattern> <startpath>\n"; "Options:"] 
            |> String.concat "\n"
        usageString

    let Usage (exitCode : int) : unit =
        let usageString = GetUsageString()
        printfn "%s\n" usageString
        Environment.Exit(exitCode)
